native/cocos/audio/apple/AudioEngine-impl.mm

```diff
@@ -0,0 +1,336 @@
+ not use Cocos Creator software for developing other software or tools that's
+ used for developing games. You are not granted to publish, distribute,
+ sublicense, and/or sell copies of Cocos Creator.
+
+ The software or tools in this License Agreement are licensed, not sold.
+ Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+****************************************************************************/
+
+#import "AVFoundation/AVAudioEngine.h"
+#import "AVFoundation/AVAudioPlayerNode.h"
+#import "AVFoundation/AVAudioFile.h"
+#import "AVFoundation/AVAudioFormat.h"
+#include "audio/apple/AudioEngine-Impl.h"
+#include "audio/include/AudioEngine.h"
+#include "application/ApplicationManager.h"
+#include "platform/FileUtils.h"
+
+static AVAudioEngine* engine_instance = nullptr;
+static AVAudioFormat* uniform_format = nullptr;
+namespace cc {
+static AudioEngineImpl* s_instance = nullptr;
+class Scheduler;
+
+
+AudioEngineImpl::AudioEngineImpl():_currentID(0), _lazyInitLoop(true){
+    engine_instance = [[AVAudioEngine alloc] init];
+    s_instance = this;
+}
+
+AudioEngineImpl::~AudioEngineImpl() {
+    if (auto sche = _scheduler.lock()) {
+        sche->unschedule("AudioEngine", this);
+    }
+    
+    //TODO: release all player nodes attached?
+    [engine_instance release];
+    _players.clear();
+    _caches.clear();
+    s_instance = nullptr;
+}
+
+bool AudioEngineImpl::init() {
+    _scheduler = CC_CURRENT_ENGINE()->getScheduler();
+    return true;
+    
+}
+
+// Return player in memo pool
+int32_t AudioEngineImpl::getUsablePlayer() {
+    if(_unusedPlayers.empty()) {
+        return -1;
+    }
+    return _unusedPlayers.begin()->first;
+}
+AudioCache* AudioEngineImpl::preload(const ccstd::string &filePath, const LoadCallback &callback) {
+    AudioCache* cache;
+    auto itr = _caches.find(filePath);
+    
+    if(itr != _caches.end()){
+        cache = itr->second;
+    } else {
+        auto fileFullPath = FileUtils::getInstance()->fullPathForFilename(filePath);
+        cache = new AudioCache(fileFullPath);
+        _caches[filePath] = cache;
+        AudioEngine::addTask([cache](){
+            cache->load();
+        });
+        _caches[filePath] = cache;
+    }
+    if(cache && callback) {
+        cache->addLoadCallback(callback);
+    }
+    return cache;
+}
+
+AudioCache* AudioEngineImpl::forceLoad(const ccstd::string &filePath, const LoadCallback &callback) {
```

Comment:
The only difference with `preload` is that `forceLoad` runs in the current thread. Can these two methods merge into one?